silkworm/infra/concurrency/context_pool.hpp

```diff
@@ -0,0 +1,185 @@
+/*
+   Copyright 2022 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#pragma once
+
+#include <atomic>
+#include <cstddef>
+#include <functional>
+#include <memory>
+#include <ostream>
+#include <vector>
+
+#include <boost/asio/io_context.hpp>
+
+#include <silkworm/infra/common/log.hpp>
+#include <silkworm/infra/concurrency/context_pool_settings.hpp>
+#include <silkworm/infra/concurrency/idle_strategy.hpp>
+
+namespace silkworm::concurrency {
+
+//! Asynchronous scheduler running an execution loop.
+class Context {
+  public:
+    explicit Context(std::size_t context_id, WaitMode wait_mode = WaitMode::blocking);
+
+    [[nodiscard]] boost::asio::io_context* io_context() const noexcept { return io_context_.get(); }
+    [[nodiscard]] WaitMode wait_mode() const noexcept { return wait_mode_; }
+    [[nodiscard]] std::size_t id() const noexcept { return context_id_; }
+
+    //! Execute the scheduler loop until stopped.
+    void execute_loop();
+
+    //! Stop the execution loop.
+    void stop();
+
+  protected:
+    //! Execute single-threaded loop until stopped.
+    template <typename WaitStrategy>
+    void execute_loop_single_threaded(WaitStrategy&& wait_strategy);
+
+    //! Execute multi-threaded loop until stopped.
+    void execute_loop_multi_threaded();
+
+    //! The unique scheduler identifier.
+    std::size_t context_id_;
+
+    //! The asio asynchronous event loop scheduler.
+    std::shared_ptr<boost::asio::io_context> io_context_;
+
+    //! The work-tracking executor that keep the asio scheduler running.
+    boost::asio::execution::any_executor<> work_;
+
+    //! The waiting mode used by execution loops during idle cycles.
+    WaitMode wait_mode_;
+};
+
+std::ostream& operator<<(std::ostream& out, const Context& c);
+
+//! Pool of \ref Context instances running as separate reactive schedulers.
+template <typename T = Context>
```

Comment:
Could we use a concrete type or interface instead of a template, and put `unique_ptr<Context>` in the contexts?